Skip non impacted product tests

nineinchnick opened the pull request:
It's a similar idea to #10323 applied to product tests, where we:
* identify impacted modules in the PR (that have any changes in them)
* map modules to plugins and extract features they provide (like connector names, password authorization etc) 
* match these impacted connectors with features (catalogs) configured in particular PT environments
* if environments don't use any of the features from impacted modules (plugins), the PTs running on these environments are skipped

Connectors used in PT environments (and other features) are defined explicitly (together when configuring catalogs) and this is verified by automatically enabling a special test that checks if all connectors defined in an env match those configured in a running Trino server.

If any non-plugin module is modified, all PTs are run. If any PT is modified, all PTs are run. Also, all PTs always run on master. 

For example, if only the `trino-postgresql` module is modified in a PR, all PT jobs will finish in ~11 minutes.

Previous attempts:
* https://github.com/trinodb/trino/pull/10730
* https://github.com/trinodb/trino/pull/10761

Fixes https://github.com/trinodb/trino/issues/1988

nineinchnick commented:
I think one product test suite should include an environment that always runs, just to have a basic smoke test proving the server can load all plugins, start correctly and handle queries, even when not querying a connector that is being modified in a PR.

bitweld commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/Environment.java:
```diff
@@ -389,6 +409,24 @@ public Builder addContainer(DockerContainer container)
 
+        public Builder addConnector(String connectorName)
+        {
+            configuredConnectors.add(connectorName);
```
check argument for null and emptiness

nineinchnick replied:
Done

bitweld commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/Environment.java:
```diff
@@ -553,6 +593,51 @@ private Environment build(Optional<EnvironmentListener> listener)
+            if (!containers.containsKey(TESTS)) {
+                return;
+            }
+            DockerContainer container = containers.get(TESTS);
```
nit: rename to testContainer and use below in String[] commandParts = testContainer.getCommandParts()

nineinchnick replied:
Done, missed it when extracting a method, thanks!

bitweld commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/Environment.java:
```diff
@@ -553,6 +593,51 @@ private Environment build(Optional<EnvironmentListener> listener)
+            String temptoConfig = "/docker/presto-product-tests/conf/tempto/tempto-configured-connectors.yaml";
+            container.withCopyFileToContainer(forHostPath(tempFile.getPath()), temptoConfig);
+            // add custom tempto config and configured_features to arguments if there are any other groups enabled
+            String[] commandParts = containers.get(TESTS).getCommandParts();
```
nit: use testContainer.getCommandParts() (earlier comment)

bitweld reviewed (commented):
great work - should bring crucial reduction of CI consumption
some initial comments, still testing code

nineinchnick commented on plugin/trino-hive-hadoop2/bin/common.sh:
```diff
@@ -211,3 +211,12 @@ function deploy_core_site_xml() {
+# Checks if Gitflow Incremental Builder (GIB) is enabled and specified module should be built
+function check_gib_impacted() {
+    local module=$1
+    if [ -f gib-impacted.log ] && ! grep -q "$module" gib-impacted.log; then
```
Please review GIB changes in #10323

nineinchnick replied:
Done in #10323

nineinchnick commented on testing/trino-plugin-reader/pom.xml:
```diff
@@ -0,0 +1,128 @@
+        <dependency>
+            <groupId>org.codehaus.plexus</groupId>
+            <artifactId>plexus-utils</artifactId>
+            <version>3.0.22</version>
```
Done

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,260 @@
+    {
+        Optional<List<String>> modules = readTrinoPlugins(rootPom);
+        if (modules.isEmpty()) {
+            return new HashMap<>();
```
Done

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,254 @@
+                        log.error(e, "Couldn't read services jar for module %s", module);
+                    }
+                    log.warn("Couldn't find plugin name in services jar for module %s", module);
+                    return null;
```
I changed this to throw an exception.

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,254 @@
+            log.error(e, "Couldn't process service JAR %s " + serviceJar);
+        }
+        log.warn("Couldn't find 'META-INF/services/io.trino.spi.Plugin' file in the service JAR %s", serviceJar);
+        return Optional.empty();
```
I changed this to throw an exception.

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,254 @@
+        }
+        catch (XmlPullParserException e) {
+            log.warn(e, "Couldn't parse file %s", modulePom);
+            return false;
```
I changed this to throw an exception.

nineinchnick commented:
There are some plugins, like `trino-password-authenticators`, that don't provide any connectors but other features, that are covered by PTs. We should extend the ideas in this PR to include other features provided by plugins, not only connectors, but we can follow the same logic - if something is configured in the test environment, then we assume it's being tested.

nineinchnick commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/Environment.java:
```diff
@@ -553,6 +642,46 @@ private Environment build(Optional<EnvironmentListener> listener)
+                                        "presto",
+                                        Map.of(
+                                                "configured_connectors", configuredConnectors,
+                                                "configured_password_authenticators", configuredPasswordAuthenticators))));
```
This is unused in tests for now.

nineinchnick commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/TestConfiguredFeatures.java:
```diff
@@ -0,0 +1,55 @@
+    private List<String> configuredConnectors;
+
+    @Inject
+    @Named("databases.presto.configured_password_authenticators")
```
This is unused here. I don't think it's possible to extract this from a running Trino server, and the only way of testing this is to actually try to log in using different authentication methods. I'll try to write a test that uses this on a fresh head.

nineinchnick replied:
I removed this, for now, as such test would rely too much on specific environment config, I'm not sure there's much value in it.

nineinchnick commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/cli/SuiteRun.java:
```diff
@@ -124,6 +126,9 @@ public Integer call()
         @Option(names = "--cli-executable", paramLabel = "<jar>", description = "Path to CLI executable " + DEFAULT_VALUE, defaultValue = "${cli.bin}")
         public File cliJar;
 
+        @Option(names = "--impacted-features", paramLabel = "<file>", description = "Skip tests not using these features " + DEFAULT_VALUE)
```
I think this change warrants creating a README for PTL, with a separate section that would describe how this works.

ksobolew reviewed (commented):
Some comments because I couldn't resist ;)

Do I understand it correctly that for each "feature" kind we need special support? Like, if there was something other than connectors or authenticators (say, a custom system access control), we would have to collect them and pass them to the environment as well?

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,253 @@
+        Map<String, String> modulesToPlugins = mapModulesToPlugins(rootPom);
+        Map<String, Plugin> plugins = loadPlugins(pluginDir).stream()
+                .map(plugin -> new SimpleEntry<>(plugin.getClass().getName(), plugin))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
Done, thanks!

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,253 @@
+                            (path, basicFileAttributes) -> path.toFile().getName().matches(".*-services.jar"))) {
+                        Optional<Path> jarFile = files.findFirst();
+                        if (jarFile.isPresent()) {
+                            return new SimpleEntry<>(module, readPluginClassName(jarFile.get().toFile()));
```
Done, thanks!

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,253 @@
+                .map(module -> {
+                    try (Stream<Path> files = java.nio.file.Files.find(Path.of(requireNonNullElse(rootPom.getParent(), ".") + "/" + module, "target"),
+                            1,
+                            (path, basicFileAttributes) -> path.toFile().getName().matches(".*-services.jar"))) {
```
👍

findepi commented:
"Avoid building tarball in every PT job" is orthogonal, right?
how much of improvement is it?
can we expedite this and merge, to see quicker whether it brings any downsides?

nineinchnick commented:
> "Avoid building tarball in every PT job" is orthogonal, right? how much of improvement is it? can we expedite this and merge, to see quicker whether it brings any downsides?

I guess. It requires adding a new job that all other PT jobs would depend on.

Building the whole project takes about 4-5 minutes. But it produces pretty large artifacts and publishing and fetching them takes about 2 minutes, so we would shave off about 3 minutes from every PT job. 

I'll create another PR with only this change.

findepi commented:
@nineinchnick fyi the build is red.

findepi commented:
@nineinchnick please confirm that **_all_** product tests will be run when

- trino-product-tests or any its dependency  (eg trino-testing-services) changes
- trino-product-tests-launcher changes 
- trino-server changes (eg provisio packaging definition)
- core changes (trino-main, trino-spi, trino-parser, or any its dependency)
- ci.yml changes

also, do we have it documented how can i _**force**_ running all product tests on a PR?
For example, should i want to diagnose a build failing on master.

nineinchnick commented:
> @nineinchnick please confirm that **_all_** product tests will be run when
>     * trino-product-tests or any its dependency  (eg trino-testing-services) changes
>     * trino-product-tests-launcher changes
>     * trino-server changes (eg provisio packaging definition)
>     * core changes (trino-main, trino-spi, trino-parser, or any its dependency)
>     * ci.yml changes
> 
> also, do we have it documented how can i _**force**_ running all product tests on a PR? For example, should i want to diagnose a build failing on master.

@findepi I can confirm that. We detect changes in product tests, launcher, `trino-testing-services`, `core/trino-server`, and `.github` in the `pt` job using the `dorny/paths-filter` action and don't pass impacted features list if they're modified. Other core or lib changes will propagate to all plugins through Maven dependencies. If there's any module modified that's not a plugin, all tests will run.

If there are changes in a plugin with features not mapped in test environments, all tests will be run.

I also added checking for `tests:all` and `product-tests:all` labels. The latter is not defined yet, you'd have to create it. 

I added a new section in the launcher's README describing this.

kokosing commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/cli/SuiteRun.java:
```diff
@@ -124,6 +126,9 @@ public Integer call()
         @Option(names = "--cli-executable", paramLabel = "<jar>", description = "Path to CLI executable " + DEFAULT_VALUE, defaultValue = "${cli.bin}")
         public File cliJar;
 
+        @Option(names = "--impacted-features", paramLabel = "<txt>", description = "Skip tests not using these features " + DEFAULT_VALUE, split = ",")
```
`Skip tests not using these features`, Please phrase it without `not`, Maybe `Run tests tests with these features`?

nineinchnick replied:
Done!

kokosing reviewed (commented):
Just skimmed.

Do you support features like group providers, authenticators (Oauth2), system access control, event listeners? Is it possible to extend a list of to define custom feature? (I haven't skimmed closely enough).

nineinchnick commented:
> Just skimmed.
> 
> Do you support features like group providers, authenticators (Oauth2), system access control, event listeners? Is it possible to extend a list of to define custom feature? (I haven't skimmed closely enough).

The plugin reader tool lists all currently existing plugin features. New ones would have to be added there, same as in PluginManager.

Test environments currently only track connectors and password authenticators. If there's an untracked feature passed in that `--impacted-features` file, all tests are run.

electrum commented on testing/trino-plugin-reader/src/test/java/io/trino/server/TestPluginReader.java:
```diff
@@ -0,0 +1,44 @@
+        PluginReader pluginReader = new PluginReader();
+        CommandLine cmd = new CommandLine(pluginReader);
+
+        StringWriter writer = new StringWriter();
```
Move above, then you can do
```java
CommandLine cmd = new CommandLine(pluginReader)
        .setOut(new PrintWriter(writer))
        .setErr(new PrintWriter(writer));

electrum commented on testing/trino-plugin-reader/src/test/java/io/trino/server/TestPluginReader.java:
```diff
@@ -0,0 +1,44 @@
+                "--impacted-modules", "src/test/resources/gib-impacted.log",
+                "--plugin-dir", "src/test/resources/server-plugins",
+                "--root-pom", "src/test/resources/pom.xml");
+        assertEquals(0, exitCode);
```
```java
assertThat(exitCode).isEqualTo(0);
assertThat(writer.toString()).isEqualTo("");

electrum commented on testing/trino-plugin-reader/src/test/resources/server-plugins/.gitignore:
```diff
@@ -0,0 +1 @@
+
```
Why do we need this empty file?

nineinchnick replied:
The test in this module makes sure the whole plugin reader runs successfully, and it points to this empty dir to read plugins from it. But I didn't want to add real JARs here that could be actually loaded as a plugin, I'm not sure how to create the smallest possible plugin JAR like this.

electrum replied:
How about changing this to a comment that explains why this file exists?

findepi replied:
that could be a README file then

nineinchnick replied:
Done, and rebased after 373 got released!

electrum commented on testing/trino-plugin-reader/pom.xml:
```diff
@@ -5,7 +5,7 @@
         <groupId>io.trino</groupId>
         <artifactId>trino-root</artifactId>
-        <version>372-SNAPSHOT</version>
+        <version>373-SNAPSHOT</version>
```
Update in previous commit

electrum commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -162,14 +167,31 @@ private static boolean isTrinoPlugin(String module)
         }
     }
 
+    private static String readPluginClassName(File rootPom, String module)
```
Move these changes into previous commit

electrum commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/cli/TestRun.java:
```diff
@@ -241,6 +256,38 @@ private Integer tryExecuteTests()
+                String[] parts = feature.split(":", 2);
+                return parts.length < 1 ? "" : parts[0];
+            }));
+            // all possible feature prefixes:
```
Change to
```java
// see PluginReader. printPluginFeatures() for all possible feature prefixes

nineinchnick commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/Environment.java:
```diff
@@ -553,6 +599,45 @@ private Environment build(Optional<EnvironmentListener> listener)
+            File tempFile;
+            try {
+                tempFile = File.createTempFile("tempto-configured-connectors-", ".yaml");
+                objectMapper.writeValue(
```
I reduced it a bit.

nineinchnick commented on pom.xml:
```diff
@@ -1437,6 +1438,12 @@
+            <dependency>
+                <groupId>org.apache.maven</groupId>
+                <artifactId>maven-model</artifactId>
+                <version>3.3.9</version>
```
I didn't check, sorry, updated.

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,247 @@
+                if (entry.isDirectory() || !entry.getName().equals("META-INF/services/io.trino.spi.Plugin")) {
+                    continue;
+                }
+                BufferedInputStream bis = new BufferedInputStream(zipFile.getInputStream(entry));
```
Done!

nineinchnick commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,247 @@
+            ZipFile zipFile = new ZipFile(serviceJar);
+            Enumeration<? extends ZipEntry> e = zipFile.entries();
+
+            while (e.hasMoreElements()) {
```
I have to admit I copied this from some examples of how to read zip files. Looks much better as a stream, thanks!

hashhar commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvMultinodeAllConnectors.java:
```diff
@@ -0,0 +1,94 @@
+    {
+        // blackhole, jmx, tpch are already configured in Standard base env
+        ImmutableList.Builder<String> listBuilder = ImmutableList.builder();
+        List<String> connectors = listBuilder
```
How can this be kept up to date?

nineinchnick replied:
I don't think we can automate this, same as we don't enforce that there are proper product tests added with new connectors.

hashhar commented on .github/workflows/ci.yml:
```diff
@@ -646,14 +646,32 @@ jobs:
           find . -type f -name \*-executable.jar -exec chmod 0777 {} \;
-      - name: Product Tests
+      - name: Product Tests (all)
+        if: steps.filter.outputs.product-tests == 'true' || contains(github.event.pull_request.labels.*.name, 'tests:all') || contains(github.event.pull_request.labels.*.name, 'tests:all-product')
+        run: |
+          testing/bin/ptl suite run \
+            --suite ${{ matrix.suite }} \
+            --config config-${{ matrix.config }} \
+            --bind=off --logs-dir logs/ --timeout 2h \
+            --trino-jdk-version zulu_${{ matrix.jdk }}
+      - name: Product Tests (impacted-features)
+        if: steps.filter.outputs.product-tests == 'false' && !contains(github.event.pull_request.labels.*.name, 'tests:all') && !contains(github.event.pull_request.labels.*.name, 'product-tests:all')
```
Labels used in these two `if` expressions are different.

nineinchnick replied:
Doh, thanks!

hashhar commented on .github/workflows/ci.yml:
```diff
@@ -646,14 +646,32 @@ jobs:
+            product-tests:
+              - 'testing/trino-product-tests*/**'
+              - 'testing/trino-testing-services/**'
+              - 'core/trino-server/**'
```
Why trino-server? Is it to catch new connectors?

nineinchnick replied:
`trino-server` is the only module that does not properly define dependencies. It relies on Provisio to find artifacts. So if anything changes in this module, GIB wouldn't properly include all plugins on its own.

hashhar replied:
Can you add a short comment that it's there because of no explicit dependencies in pom?

nineinchnick replied:
Done!

hashhar commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/cli/SuiteRun.java:
```diff
@@ -124,6 +126,9 @@ public Integer call()
         @Option(names = "--cli-executable", paramLabel = "<jar>", description = "Path to CLI executable " + DEFAULT_VALUE, defaultValue = "${cli.bin}")
         public File cliJar;
 
+        @Option(names = "--impacted-features", paramLabel = "<file>", description = "Only run tests in environments with these features " + DEFAULT_VALUE)
```
Would be nice to be able to provide inline input for making use of this locally without having to create a file (or can we use bash process substitution magic for that)?

nineinchnick replied:
Yes, process substitution works here. The list of impacted features will be very long, that's why we went with passing it as a file, and I doubt it'll be used when executing this manually very often.

hashhar commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/Environment.java:
```diff
@@ -389,6 +419,56 @@ public Builder addContainer(DockerContainer container)
             return this;
         }
 
+        public Builder addConnector(String connectorName)
```
nit: `addCatalog` instead of `addConnector`.

BTW if I added a catalog without using this method `selectConfiguredConnectors` should fail right?
But not if I added a password authenticator without this method?

nineinchnick replied:
I have a little bit of a problem with this, because this method does exactly this - it only registers a connector being used, it does not copy any catalog configs. But the other variants of it do. Can I keep this one and rename others and explain that in a comment?

Regarding password authenticators:
Because currently there's nothing in Trino that would report all configured password authenticators, writing a test that would verify them is pretty hard because it would have to assume some common way of authenticating using every possible way (password, oauth).

Maybe we could add a system table to report this, something like: https://github.com/trinodb/trino/pull/10830

We do pass a list of password authenticators configured in test environments to Tempto, but there's no test to use it: https://github.com/trinodb/trino/pull/10895#discussion_r803904845

hashhar replied:
Let's ignore my comment for the simplicity of overloaded methods.

For password authenticators (and future features) we may end up just checking at review time like we do today when adding a new connector.

hashhar reviewed (commented):
Some questions mostly.

Skimmed "[Add plugin reader utility class](https://github.com/trinodb/trino/pull/10895/commits/2636246c9a4bda905f2fead0bdd65e142a587d87)"  - reviewed rest.

hashhar commented on testing/trino-product-tests/README.md:
```diff
@@ -244,6 +244,32 @@ You can now start the debug configuration that you just created and IntelliJ
+A list of modules modified on a particular git branch can be obtained by enabling the `gib` profile
+when building Trino. It's saved as `gib-impacted.log`.
+
+> Note: all product tests should be run when there are changes in any commont files, like:
```
```suggestion
> Note: all product tests should be run when there are changes in any common files, like:
```

nineinchnick replied:
Done!

hashhar reviewed (approved):
LGTM from me.

PTAL @trinodb/maintainers so that we can get this in.

hashhar commented on testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java:
```diff
@@ -0,0 +1,255 @@
+
+    private static void printPluginFeatures(Plugin plugin)
+    {
+        plugin.getConnectorFactories().forEach(factory -> System.out.println("connector:" + factory.getName()));
```
Is it possible to add test that would detect when new methods have been added to `Plugin` via reflection?

(Not required but useful if we want to maintain this class).

kokosing commented:
> PTAL https://github.com/orgs/trinodb/teams/maintainers so that we can get this in.

👍 Let's move forward with this. It is invasive change for CI anyway, so we need to have eyes wide open for a while for any CI issues.

electrum commented on testing/trino-plugin-reader/src/test/resources/server-plugins/README.md:
```diff
@@ -0,0 +1,4 @@
+This is an empty directory because the plugin reader needs this to exist,
+so it's test won't fail.
```
Grammar: its

electrum commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvMultinodeAllConnectors.java:
```diff
@@ -0,0 +1,94 @@
+                .add("bigquery")
+                .add("cassandra")
+                .add("clickhouse")
+                .add("druid")
```
Missing `delta-lake`

nineinchnick replied:
Done! Rebased on latest master (Trino 375)

findepi replied:
How can we ensure this stays in sync? We'll forget to add things here

nineinchnick replied:
Ashhar also asked about that: https://github.com/trinodb/trino/pull/10895#discussion_r830948904

electrum commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvMultinodeAllConnectors.java:
```diff
@@ -0,0 +1,94 @@
+    public void extendEnvironment(Environment.Builder builder)
+    {
+        // blackhole, jmx, tpch are already configured in Standard base env
+        ImmutableList.Builder<String> listBuilder = ImmutableList.builder();
```
Inline this

nineinchnick replied:
Done! For some reason, I thought that `List.of()` only supports up to 10 items and we had more here.

electrum commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvMultinodeAllConnectors.java:
```diff
@@ -0,0 +1,94 @@
+    public EnvMultinodeAllConnectors(StandardMultinode standardMultinode, DockerFiles dockerFiles)
+    {
+        super(standardMultinode);
+        this.configDir = requireNonNull(dockerFiles, "dockerFiles is null").getDockerFilesHostDirectory("conf/environment/multinode-all/");
```
No need for null check here, since it will NPE directly on access.

ksobolew replied:
I guess we can remove all those once we standardize on Java 17 (improved NPE messages were released in Java 14)

electrum commented on .github/workflows/ci.yml:
```diff
@@ -662,6 +662,9 @@ jobs:
               - 'testing/trino-product-tests*/**'
               - 'testing/trino-testing-services/**'
+              # run all tests when there are any changes in the trino-server Maven module
+              # because it doesn't define it's Trino dependencies and
```
Grammar: its

electrum replied:
This part is redundant. We could say

> because it uses Provisio to determine its dependencies

electrum commented on testing/trino-product-tests/README.md:
```diff
@@ -244,6 +244,32 @@ You can now start the debug configuration that you just created and IntelliJ
+Test environments track which Trino features they enable, like connectors or password authenticators.
+This can be used to skip running product tests on environments that don't use specific features, by passing
+`--impacted-features=<file>` to the product test launcher. The file should contain a list of features to test,
+one per line, as `feature-kind:feature-name`, ex. `connector:hive`.
```
ex. -> e.g.,

electrum commented on testing/trino-product-tests/README.md:
```diff
@@ -244,6 +244,32 @@ You can now start the debug configuration that you just created and IntelliJ
+`--impacted-features=<file>` to the product test launcher. The file should contain a list of features to test,
+one per line, as `feature-kind:feature-name`, ex. `connector:hive`.
+
+Such a file can be generated from a list of Maven modules that are Trino plugins,
```
Remove comma

electrum commented on testing/trino-product-tests/README.md:
````diff
@@ -244,6 +244,32 @@ You can now start the debug configuration that you just created and IntelliJ
+    -p core/trino-server/target/trino-server-*-hardlinks/plugin
+```
+
+A list of modules modified on a particular git branch can be obtained by enabling the `gib` profile
````
git -> Git

electrum commented on testing/trino-product-tests/README.md:
```diff
@@ -244,6 +244,32 @@ You can now start the debug configuration that you just created and IntelliJ
+A list of modules modified on a particular git branch can be obtained by enabling the `gib` profile
+when building Trino. It's saved as `gib-impacted.log`.
+
+> Note: all product tests should be run when there are changes in any common files, like:
```
like: -> including

nineinchnick replied:
All done!

MiguelWeezardo commented:
@hashhar could you merge this? Two approvals should be enough.

hashhar reviewed (approved):
I'll merge once CI is done, re-reviewed and rebased to ensure CI results are fresh.

martint commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/cli/TestRun.java:
```diff
@@ -219,19 +244,57 @@ public Integer call()
         }
 
-        private Environment startEnvironment()
+        private boolean hasImpactedFeatures(Environment environment)
```
What's a "feature" in this context? How do we track them, keep them up to date, etc?

nineinchnick replied:
Please read the PR description that explains this and also previous comments, like https://github.com/trinodb/trino/pull/10895#issuecomment-1059520957

martint replied:
I saw:

> map modules to plugins and extract features they provide (like connector names, password authorization etc)

But I still don't understand how we maintain that mapping.

nineinchnick replied:
The plugin interface has a method for each feature. I wrote a plugin reader tool that lists every feature provided by every plugin. It's similar to what the plugin manager does. If there would be a new kind of feature added to this interface, it should be handled in both the plugin manager and plugin reader.

nineinchnick replied:
@Ashhar suggested adding a test that would use reflection to make sure all plugin features are covered in the plugin reader. Do you think adding such a test is required to merge this?
